gpjax/natural_gradients.py

```diff
@@ -0,0 +1,288 @@
+    m = natural_vector.shape[0]
+
+    # S⁻¹ = -2θ₂
+    S_inv = -2 * natural_matrix
+    S_inv += I(m) * jitter
+
+    # S⁻¹ = LLᵀ
+    L = jnp.linalg.cholesky(S_inv)
+
+    # C = L⁻¹I
+    C = jsp.linalg.solve_triangular(L, I(m), lower=True)
+
+    # S = CᵀC
+    S = jnp.matmul(C.T, C)
+
+    # μ = Sθ₁
+    mu = jnp.matmul(S, natural_vector)
+
+    # η₁ = μ
+    expectation_vector = mu
+
+    # η₂ = S + μ μᵀ
+    expectation_matrix = S + jnp.matmul(mu, mu.T)
+
+    return {
+        "expectation_vector": expectation_vector,
+        "expectation_matrix": expectation_matrix,
+    }
+
+
+def _expectation_elbo(
+    posterior: AbstractPosterior,
+    variational_family: AbstractVariationalFamily,
+    train_data: Dataset,
+) -> tp.Callable[[dict, Dataset], float]:
+    """
+    Construct evidence lower bound (ELBO) for varational Gaussian under the expectation parameterisation.
+    Args:
+        posterior: An instance of AbstractPosterior.
+        variational_family: An instance of AbstractVariationalFamily.
+    Returns:
+        Callable: A function that computes ELBO.
+    """
+    evg = ExpectationVariationalGaussian(
+        prior=variational_family.prior,
+        inducing_inputs=variational_family.inducing_inputs,
+    )
+    svgp = StochasticVI(posterior=posterior, variational_family=evg)
+    identity_transformation = build_identity(svgp._initialise_params(jr.PRNGKey(123)))
+
+    return svgp.elbo(train_data, identity_transformation, negative=True)
+
+
+def _stop_gradients_nonmoments(params: tp.Dict) -> tp.Dict:
+    """
+    Stops gradients for non-moment parameters.
+    Args:
+        params: A dictionary of parameters.
+    Returns:
+        tp.Dict: A dictionary of parameters with stopped gradients.
+    """
+    trainables = build_trainables_false(params)
+    moment_trainables = build_trainables_true(params["variational_family"]["moments"])
+    trainables["variational_family"]["moments"] = moment_trainables
+    params = trainable_params(params, trainables)
+    return params
+
+
+def _stop_gradients_moments(params: tp.Dict) -> tp.Dict:
+    """
+    Stops gradients for moment parameters.
+    Args:
+        params: A dictionary of parameters.
+    Returns:
+        tp.Dict: A dictionary of parameters with stopped gradients.
+    """
+    trainables = build_trainables_true(params)
+    moment_trainables = build_trainables_false(params["variational_family"]["moments"])
+    trainables["variational_family"]["moments"] = moment_trainables
+    params = trainable_params(params, trainables)
+    return params
+
+
+def natural_gradients(
+    stochastic_vi: StochasticVI,
+    train_data: Dataset,
+    transformations: dict,
+    # bijector = tp.Optional[dx.Bijector] = Identity, #bijector: A bijector to convert between the user chosen parameterisation and the natural parameters.
+) -> tp.Tuple[tp.Callable[[dict, Dataset], dict]]:
+    """
+    Computes natural gradients for variational Gaussian.
```

Comment:
Be explicit about what the gradient is being computed with respect to i.e., the natural parameters.